## **1.python comment**

In [ ]:
# single line comment ## ctrl + ?
""" this is a 
multiple line 
comment """

' this is a \nmultiple line\ncomment '

In [ ]:
#Python Keywords ## shift+alt+b
"""  95% lower case
 5% uppercase
 python is case senstive """

# 3 keyword Uppercase #askey value
True,False,None



(True, False, None)

## **2.data types**

### **python variables** 

Rules:

* starts with Alphabets or underscore **(a-z,A-z)
* then can he a combination of Alphabets 
* cannot be keywords
* no speacial char not allowted 
* only allows speacial  char _
* after alphabets can add numbers
* number not allowed in first 


In [3]:
#integer

name ="david"
age = 27

david_craj =  "GenAi developer"

a = 100
A = 200
a = 400

_and  = "chinna"

A10 = 1000

____ = "Honey"

# a,b, *c = "hai","hello","Welcome","genAi" #value speatrators *Unpacked
 
a=b=c=5000

c =5000
c=b
b=a




### Data Types
* interger -1,2
* Float -0.2
* complex -1+0j(j represents that its complex)
* string -"hello world"
* Boolean - True,False


In [4]:
#integer 1,-2 - Numeric value

a = 500
a1 = -500


#Float - 1.0, 1e10 ,  - Decimal value , e-exhansion

b = 50.55
b1 = -50.55


b2 = 2e5


# Complex  - imaginary numbers exam
# any number last adding j called complex

c =20j


#string 

# string - "hello", "world"

d =  "hai"
d1 = 100 + 50


#Boolean  

e = True
e1 =  False




#### print() - Python standared functions
not stores in memory


In [5]:
print(100)

100


#### **escape Sequences**

to handle string we are using escape sequence
* \ - for Newline in  editior
* \\\ - prints \
* \' - prints '
* \n -prints a new line
* \t -prints a tab


In [6]:
print("hello\\ world\\")
print("10\\100")
print("'I\' am python")
print("hello\"world\"")
print("hello\nworld")
print("""Lorem Ipsum is simply dummy text of the printing 
        and typesetting industry. Lorem Ipsum has been the 
        industry's standard dummy text ever since 1966, when  
         at Letraset and James Mosley, the librarian at  
         St Bride Printing Library in London, took a 1914 Cicero
         translation and scrambled it to make dummy text for Letraset's Body Type sheets. It has survived not only many decades, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised thanks to these sheets and more recently with desktop publishing software like Aldus PageMaker and Microsoft Word including versions of Lorem Ipsum.""")

hello\ world\
10\100
'I' am python
hello"world"
hello
world
Lorem Ipsum is simply dummy text of the printing 
        and typesetting industry. Lorem Ipsum has been the 
        industry's standard dummy text ever since 1966, when  
         at Letraset and James Mosley, the librarian at  
         St Bride Printing Library in London, took a 1914 Cicero
         translation and scrambled it to make dummy text for Letraset's Body Type sheets. It has survived not only many decades, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised thanks to these sheets and more recently with desktop publishing software like Aldus PageMaker and Microsoft Word including versions of Lorem Ipsum.


### ****ascli key or number (ascli table)**
in anykeword in keyborad takes a  number
python intepreater program language
ord() - function #chr to number 
chr() - function #num to chr

In [7]:
print("David")
print(chr(68))
print(ord("a"))
print(ord("v"))
print(ord("i"))
print(ord("d"))

David
D
97
118
105
100


In [80]:
#end , sep 
#sep
#The symbol ** is read as double **asterisk or exponentiation operator.
#\n -new line
# \t - tab

print(1,2,3,4,5,sep="\n")
print(6,7,8,9,10)

1
2
3
4
5
6 7 8 9 10


In [82]:
#end 
 # end="\n" → new line
 # end=" "  → space
 # end="!"  → !

print(1,2,3,4,5,end=" ")

print(6,7,8,9,10,end=" ")
print(11,12,13,14,15)

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15


In [ ]:
# num1 = int(input("Enter a number Num1: "))
# num2 = int(input("Enter a number Num2: "))

# total = num1+num2
# print(total)

12


##  **Trio : value , id, type**
 Id -  address for memory location can access
 en numbers of memory loaction
 in our laptop memory laction saves
 useful for mutable and ubmutable

Each variable has three types: a value, ID, and its type. We can access them by using built-in functions named id() and type(). Normally, the value is accessed easily, but to know its ID and type, these functions are used.


In [11]:
a = 100
print(a)
print(type(a))
print(id(a))

100
<class 'int'>
140714084856024


### ***TYPE CASTING***
* python standared function

There may be times when you want to specify a type onto a variable. This can be done with casting. In simplified terms, one data type can be converted to another data type.

#Functions for Type Casting
* int()
* float()
* str()
* bool()
* list()
* tuple()
* set()

In [12]:
#int -  we can Typecasting to ,#string,#float,#boolan,#complex
# numberic data types can change to numberic #ex  num ="500"
# each saves in differnt memory location

a = 500
a1  = float(a)
a2  = str(a)
a3  = bool(a)
# print(a , type(a),id(a))
# print(a1 , type(a1),id(a1))
# print(a2 , type(a2),id(a2))
# print(a3 , type(a3),id(a3))

#boolean - -  we can Typecasting to ,#string,#float,#boolan,#complex
# each saves in differnt memory location

b = 40.55
b1  = int(b)
b2  = str(b)
b3  = bool(b)
# print(b, type(b))
# print(b1,type(b1))
# print(b2,type(b2))
# print(b3,type(b3))

#str -  can change only boolean data type only #bool
# each saves in differnt memory location
# char cannot change as Number

c ="hello"
# c1 = int(c)
# c2 = float(c)
c3 = bool(c)
# print(c,type(c))
# print(c3,type(c3))


d = "500"
d1 = int(d)
d2 = float(d)
d3 = bool(d)
# print(d,type(d))
# print(d1,type(d1))
# print(d2,type(d2))
# print(d3,type(d3))

#bool - 

e = 10
e1 = bool("")  # bool => False
e2 = bool(0)   # bool => False
# print(e,type(e))
# print(e1,type(e1))
# print(e2,type(e2))




### print() & input()
### input( ) - function

input() - always recevies a string type , we have type  caste

print() function has by default a newline and white space in it.
input() function can also be used to print string and get input.

In [ ]:
num = 100
# print("the value of num = ", num)

the value of num =  100


In [32]:

num1 = input("Enter a Number: ")
x = int(num1)
print("The value of Num = ",x,type(x))

num2 = int(input("Enter a Number: "))
print("The value of Num = ",num2,type(num2))

total = x + num2
print("sum of total ", total)

The value of Num =  200 <class 'int'>
The value of Num =  300 <class 'int'>
sum of total  500


#### **del keyword**
deleting unwanted varibles thats stores in memory location
unwanted memory location runs in code while excusting
#deleting unwanted variables in memory loacton
#garbage 

In [74]:

# num1 = input("Enter a Number: ")
# x = int(num1)
# del num1  # i have removed  num1 from memory location
# print("The value of x = ",x,type(x),id(x))
# print("The value of Num = ",num1,type(num1),id(num1))


## **3.operators**
A mathamatical functions operation called operators

Arithmetic operters - +Addition , -substractions , *multification , /Division, **powerof, //Quotient , %Remainder



| Operator Types       | Operands                                        |
| -------------------- | ----------------------------------------------- |
| Arithmetic operators | `+`, `-`, `*`, `/`, `**`, `//`, `%`             |
| Assignment operators | `=`, `+=`, `-=`, `/=`, `*=`, `**=`, `//=`, `%=` |
| Comparison operators | `==`, `!=`, `>`, `<`, `<=`, `>=`                |
| Logical operators    | `and`, `or`, `not`                              |
| Identity operators   | `is`, `is not`                                  |
| Membership operators | `in`, `not in`                                  |


In [ ]:
###Arithmetic operters -
#  +Addition , -substractions , *multification ,
#  /Division, **powerof, //Quotient , %Remainder

# print(10+3)  # Add - 13
# print(10-3)  # sub - 7
# print(10*3)  # multi - 30
# print(10/3)  # div - 3.3333... 
# print(10**3) # pwo = 10*10*10 = 1000
# print(10//3) # Q = 3
# print(10%3)  # R = 1


13
7
30
3.3333333333333335
1000
3
1


###  **** | Assignment operators | `=`, `+=`, `-=`, `/=`, `*=`, `**=`, `//=`, `%=` ****
#### value assigning called assignment operators


In [ ]:

a = 10
print(a)

a +=5 # a = a(10) + 5
print(a)

a *=2 # a = a(15) * 2
print(a)

c =0
c +=20
print(c)

10
15
30
20


### **comparison operators**
### == , !=, >,<,>=,<=,



In [1]:
print(10==10)
print(10!=10)
print(10>10)
print(10>=10)
print(10<10)
print(10<=10)

True
False
False
True
False
True


### **logical opertors**
#and ,or,Not

In [ ]:
print('--------And-------Mutiplication')

print(10==10 and 6>4) # T and T = true
print(10!=10 and 6>4) # F and T = False 
print(10==10 and 6<4) # T and F = False
print(10!=10 and 6<4) # F and F = False

print('--------OR-------Addition')

print(10==10 or 6>4) # T or T = True
print(10!=10 or 6>4) # F or T = True
print(10==10 or 6<4) # T or F = True
print(10!=10 or 6<4) # F or F = False

print('--------NOT------- ')

print(not 10 == 10) #False
print(not 10!=10)   #True



--------And-------
True
False
False
False
--------OR-------
True
True
True
False


## **4.Python Collections**

* List is a collection which is ordered and changeable. Allows duplicate members. Denoted by []

* Tuple is a collection which is ordered and unchangeable. Allows duplicate members. Denoted by ()

* Set is a collection which is unordered and unindexed. No duplicate members. Denoted by {}

* Dictionary is a collection that is unordered, changeable (mutable), and stores data in key-value pairs. Keys must be unique (no duplicate keys). It is denoted by curly braces {} using the format {key: value}.
Note: Python nested collections are also possible.

#### **MUTABLE VS IMMUTABLE OBJECTS**

* Objects of built-in types like (int, float, bool, str, tuple) are immutable.
* Objects of built-in types like (list, set, dict) are mutable.

### * Immutable
Same data stores in Same location
Different Data stores in Different location

### * Mutable
Same data stores in Different location
Different Data stores in Different location

In [14]:
#List  - Ordered,Changeable,Duplicates,
#mutable Type ()

#### * Immutable
# Same data stores in Same location
# Different Data stores in Different location

myList1 = ["david", 10, False,20.45,"chinna"]
myList2= ["david", 10, False,20.45,"chinna"]
print(myList1)

print(myList1,id(myList1))
print(myList1,id(myList2))


['david', 10, False, 20.45, 'chinna']
['david', 10, False, 20.45, 'chinna'] 2920477483456
['david', 10, False, 20.45, 'chinna'] 2920473637184


### **python Collection**
**List** is a collection which is ordered and changeable. Allows duplicate members. Denoted by **Square Bracket[]**


In [33]:
  #         0 1 2   3      4       5
mylist = [10,20.55,"hello",False,"hello"]
print(mylist,type(mylist))

print(mylist[4])

mylist[0] = "welcome"
print(mylist)

mylist.append("david")
print(mylist)

[10, 20.55, 'hello', False, 'hello'] <class 'list'>
hello
['welcome', 20.55, 'hello', False, 'hello']
['welcome', 20.55, 'hello', False, 'hello', 'david']


#### ***list in collections methods**

In [ ]:
# Creating lists for different apps

playlist = ["Shape of You", "Naa Ready", "Believer", "Tum Hi Ho"]

favourite_foods = ["Pizza", "Burger", "Dosa", "Biryani"]

recent_locations = ["Home", "Airport", "Work", "Mall"]

print("Spotify Playlist:", playlist)
print("Zomato Foods:", favourite_foods)
print("Uber Locations:", recent_locations)


#list slicing

print("top 2 songs:",playlist[0:2])

print("last 2 loactions:",recent_locations[-2:])


## list methods

# append 

playlist.append("en thedal nee")
print("after append: ",playlist)

#insert - can add in any index between

playlist.insert(3,"song_david")
print("after insert: ",playlist)

#remove 
playlist.remove("Naa Ready")
print("after remove:",playlist)

#pop  - removes last index 
playlist.pop()
print("after pop: ",playlist)

#reverse
playlist.reverse()
print("after reverse: ",playlist)

#count
print(playlist.count("song_david"))


In [ ]:
# Creating lists for different apps

playlist = ["Shape of You", "Naa Ready", "Believer", "Tum Hi Ho"]

favourite_foods = ["Pizza", "Burger", "Dosa", "Biryani"]

recent_locations = ["Home", "Airport", "Work", "Mall"]

artist_name = ["david","yesu","sam","mahi"]

# print("Spotify Playlist:", playlist)
# print("Zomato Foods:", favourite_foods)
# print("Uber Locations:", recent_locations)


#list slicing

# print("top 2 songs:",playlist[-3:])

#list iteration

# for food in favourite_foods:
#     print("all foods",food)

# zip()  - → 1st song + 1st artist, 2nd song + 2nd artist, etc.

# for song,art in zip(playlist,artist_name):
#     print(song + " by",art)


#check if

# if "Dosa" in favourite_foods:
#     print("yes")

# favourite_foods[1] = "sharwama"
# print(favourite_foods)

# mixed  = ["david",27,1.78]

# for a in mixed:
#     print(a)

# enumerate () -  print both index and value

# for i,location in enumerate(recent_locations):
#     print(f"location {i}:{location}")

###  **Tuple**  is a collection which is ordered and unchangeable. Allows duplicate members. Denoted by **() parentheses**


In [41]:
mytuble =(10,20.55,"hello",False,"hello")
print(mytuble,type(mytuble))

print(mytuble[2])
# mytuble[3] = "Python"
# print(mytuble)

(10, 20.55, 'hello', False, 'hello') <class 'tuple'>
hello


In [ ]:
trip_summary = ("ubergo","chennai","airport","450.00","completed")
print(trip_summary)

# print(trip_summary[1])

# for item in trip_summary:
#     print(item)

# print(len(trip_summary))
print(trip_summary.count("completed"))
print(trip_summary.index("airport"))

###  **Set**

### **Set** is a collection which is unordered and unindexed. No duplicate members. Denoted by **{}- curly brackets**



In [39]:
myset = {}
print(myset,type( myset))

{} <class 'dict'>


In [40]:
mylist = []
print(mylist,type(mylist))

[] <class 'list'>


### **Dictionary**
Dictionary is a collection that is unordered, changeable (mutable), and stores data in key-value pairs. Keys must be unique (no duplicate keys). It is denoted by curly braces {} using the format **{key: value}**.
Note: Python nested collections are also possible.

In [7]:
mydict = {
    "name" : "david",
    "course":"python",
    "age": 27,
    "name":"yesu"
}
print(mydict,type(mydict))
mydict["age"]=28
mydict["city"] ="chennai"
print(mydict)


{'name': 'yesu', 'course': 'python', 'age': 27} <class 'dict'>
{'name': 'yesu', 'course': 'python', 'age': 28, 'city': 'chennai'}


In [ ]:
trip = {
    "trip_id":"UB12345",
    "pickup" : "chenai central",
    "drop" : ["Airport","Tambram","medavakam"],
    "fare" : 430.75,
    "diver":"ram",
    "status":"completed"

}

# print(trip["pickup"])
# print(trip.get("Airport"))

# print(trip.keys())
# print(trip.values())


# for key,value in trip.items():
#     print(key, ":" ,value)

# trip.update({"clear_model":"suzuki"})
# trip.update({"clear_model":"abc"})
# print(trip)

# trip.pop("status")
# print(trip)
# print(trip["drop"][1])


for location  in trip["drop"]:
    print(location)

In [ ]:
trips = {
    "UB001":{"trip_id": "UB001", "pickup": "Chennai", "drop": "Airport", "fare": 430},
    "UB002":{"trip_id": "UB002", "pickup": "Tambaram", "drop": "Chennai", "fare": 320},
    "UB003":{"trip_id": "UB003", "pickup": "T-Nagar", "drop": "Velachery", "fare": 210}
}

for trip_id ,details in  trips.items():
        print(trip_id)
        print(details["pickup"], "-->" ,details["drop"])

### **IMMUTABLE OBJECTS**

* Objects of built-in types like (int, float, bool, str, tuple) are immutable.

### * Immutable - Unchangable
* Same data stores in Same location
* Different Data stores in Different location



In [12]:
a = 100
b = 100
c = 101
print(a,type(a),id(a))
print(b,type(b),id(b))
print(c,type(c),id(c))

100 <class 'int'> 140718926786776
100 <class 'int'> 140718926786776
101 <class 'int'> 140718926786808


In [19]:
x =(10,20,30,40,50)
y =(10,20,30,40,50)
print(x,type(x),id(x))
print(y,type(y),id(y))

(10, 20, 30, 40, 50) <class 'tuple'> 2173652938432
(10, 20, 30, 40, 50) <class 'tuple'> 2173651855840


### **MUTABLE OBJECTS**

* Objects of built-in types like (list, set, dict) are mutable.

### * Mutable - changable
* Same data stores in Different location
* Different Data stores in Different location

In [20]:
x =[10,20,30,40,50]
y =[10,20,30,40,50]

print(x,type(x),id(x))
print(y,type(y),id(y))

(10, 20, 30, 40, 50) <class 'tuple'> 2173653713824
(10, 20, 30, 40, 50) <class 'tuple'> 2173652938432


s### **identity opertors**  (True or False)
* #### **is , is not**


In [23]:
## identity operators (True or False)
# is,is not

a = 100
b = 100

print(a!=b)  # "=="  =>value True
print(a is not b) # "is" =>location => True


False
False


In [26]:
x = [10,20,30,40,50]
y = [10,20,30,40,60]

print(x == y) 
print(x is y)

False
True


In [ ]:
x = (10,20,30,40,50)
y = (10,20,30,40,50)

print(x != y) 
print(x is y)

## **operators**



#### *memership operators*

* **in ,not in**

#### *memership operators*
* in ,not in
* is  - check is location
* not is - check wheather is there or not


In [2]:

print("e" in "hello world")

x = [10,20.55,"hello world"] # takes item by item
# print('e' is x)


True


##  **5.if ....elif**

IF ... ELSE

Python uses if else to choose between choices. It is called a Decision Statement.

Keywords
if
else
elif

### ***#if ,elif, else - Decision statement***

* indentataion press **tab**
* **pass key** - stop code executions
* synatx
* **if \<condition>\:**
* **elif  \<condition>\:**
* **else:**


In [ ]:
a = 101

if a > 100:
  print(a,"a is bigger than 100")

print("this lines Always works")

    

101 a is bigger than b
this lines Always works


In [21]:
a = 10

if a > 100:
  print(a,"a is bigger than 100")

else:
  print(a,"a smaller than 100")

print("this lines Always works")



10 a smaller than 100
this lines Always works


In [26]:
age = int(input("Enter your age"))

if age> 30:
    print("Age is 40")
else:
    print("Age is under 30")

Age is under 30


In [32]:
a = 100

if a > 100:
  print(a,"a is bigger than 100")

elif a == 100:
  print(a,"is equal to 100 ")

else:
  print(a,"a smaller than 100")

print("this lines Always works")


100 is equal to 100 
this lines Always works


### **if , elif ,else rules**
* **if(only once per code block(Mandtory))**
* **elif(as many time as per code block(optional))**
* **else(only once per code block(optional))**

In [41]:
a = 25

if a>50 :
   print("a is greater ")

elif a == 20 :
    print("a is 20")

elif a == 30 : 
    print("a is 30")

elif a == 40 : 
    print("a is 40")

else:
    print("a  is smaller")

a  is smaller


### **Nested if & else**

In [44]:
a = -5

if a>10:
    print("a is bigger")
    if a>100:
        print("a is biggest")
    else:
        print("a is btw 10 to 99")
elif a<10:
    print("a is smaller")
    if a < 0:
        print("a is nagative")
    else:
        print("a is btw 0 to 9")
print("code over")

a is smaller
a is nagative
code over


### **short hand if & else**

In [53]:
a = 100
b = 500

if a>b:print("a is bigger than b")
else: print("b is bigger than a")
    

b is bigger than a


In [54]:
a = 100
b = 50

print("a is bigger than b")  if a>b else print("b  is bigger than a") 


a is bigger than b


In [ ]:

#game guess project

score = 100
secret_number = 50
guess = int(input("Enter a guess  number: "))


if guess == secret_number :
    print("correct!")
    score = score + 20
    print("score is equal ",score)

elif guess > secret_number:
    print("Too high!")
    score = score - 10
    print("score is ",score)

else:
   print("too low!")
   score =  score - 10
   print("your score is :",score)

Too high
score is  90


### **6.Python While loop** - *Condition statements*

*syntax*
**while \<condition>:**
* **crl+shift+alt+enter** - "Debug"




In [11]:
a = 1  # initialization / starting point

while a <=10: # comdition / Execution
    print(a, end = " ") # print all values one line
    a += 1  # incremention 

print("\nLoop is over") # string escape sequence \n - new line
    



1 2 3 4 5 6 7 8 9 10 
Loop is over


In [22]:
a = 100

while a >=0: 
    print(a, end = " ") 
    a -= 10 

print("\nLoop is over") 
    

100 90 80 70 60 50 40 30 20 10 0 
Loop is over


In [3]:
a = 1

while a<=10: 
    print(a, end = " ") 
    a += 1 

print("\nLoop is over") 

1 2 3 4 5 6 7 8 9 10 
Loop is over


### **break**- *stop the loop*

In [9]:
# a = 1 

# while a>0:
#   print(a,end=" ")
#   a += 1

In [2]:
a = 1

while a>0:
    if a == 5:
        print(a)
        break # code ends
    print(a,end=" ")
    a += 1

1 2 3 4 5


In [30]:
list = [1,2,3,4,5]

for each in list:
    if each == 4:
        break
    print(each, end =" ")


1 2 3 

#### **continue** *- go back to the top of my loop*

In [ ]:
a = 1

while a<=20:
    if a%3 == 0:
        print(a,end=" ")
        a+= 1
        continue
    
    a += 1

print("\nloop over")

3 6 9 12 15 18 
loop over


### Task -1
### **Nested loop while loop**

In [ ]:
12345
12345
12345
12345
12345

In [10]:
a = 1

while a<=5:
    if a == 5:
        print(a,end="")
        a += 1
        continue
    
    print(a,end=" ")
    a += 1

       

1 2 3 4 5

In [58]:
a = 1

while a<=5:
    b = 1
    while b<=5:
      print(b,end="")
      b += 1
    a += 1
    print()

12345
12345
12345
12345
12345


In [3]:
a = 1

while a<=3:
    b = 1
    while b<=3:
      print(b,end="")
      b += 1
    a += 1
    print()

123
123
123


In [ ]:
a = 1

while a<=5:
    b = 1
    while b<=5:
      print(b,end="")
      b += 1
    a += 1
    print()

12345
12345
12345
12345
12345


In [1]:
## Chakaravathy

a = int(input())
b = int(input())

x = 1

while x<=a:
    y=1
    while y<=b:
        print(y,end="")
        y += 1

    x += 1
    print()



12
12
12
12
12


In [3]:
## vekadesh

row = int(input())
column = int(input())

a = 1

while a<=row:
    b = 1
    while b<=column:
        print(b,end="")
        b += 1

    a += 1
    print()
            


12
12
12
12
12


In [4]:
## Baby

row = int(input())
column = int(input())

i = 1

while i<=row:
    j = 1
    while j<=column:
        print(j,end="")
        j += 1

    i += 1
    print()



1234
1234
1234
1234
1234
1234


In [5]:
## Faizal


row = int(input())
column = int(input())

a = 1

while a<=row:
    b = 1
    while b<=column:
        print(b,end="")
        b += 1

    a += 1
    print()



12345
12345
12345


In [6]:
## David

row = int(input())
column = int(input())

i = 1

while row>=i:
    j = 1
    while column>=j:
        print(j,end="")
        j += 1

    i += 1
    print()


1234567
1234567
1234567
1234567


In [9]:
value = input("Enter True or False: ")

if value == "False":
    print("Hello")
elif value == "True":
    print("Hi")
else:
    print("Bye")

Bye


### **python for loop**

For loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string). With the for loop we can execute a set of statements, once for each item in a list, tuple, set etc.

Keyword

### **for**

Note: range() – is a Python standard function used for numbers.

### syntax

* **for \<variable_name>\ in <str/list/tuple/set/dict>:**


In [13]:
for x in "hello world":
    print(x,end="")

hello world

In [33]:
mystr = "hello world" ## 

for x in mystr: ## end before 0:5 ## fetechs letter by letter
    print(x,end="")

hello world

In [42]:
## list collections

mylist = [10,20.55,"hello",True,20.55,"hello"] ## Items

for x in mylist:
    print(x,end=",")

10,20.55,hello,True,20.55,hello,

In [47]:
## tuple collections

mylist = (10,20.55,"hello",True,20.55,"hello")

for x in mylist:
    print(x,end=",")

10,20.55,hello,True,20.55,hello,

In [48]:
## set collections

mylist = {10,20.55,"hello",True,20.55,"hello"} 

for x in mylist:
    print(x,end=",")

True,10,20.55,hello,

In [63]:
## dict collections

mylist = {
    "name":"david",
     "age": 27,
      "city":"chennai",
      } 

for key,value in mylist.items():
    print(key,":",value)

name : david
age : 27
city : chennai


In [73]:
## string we cannot call in range

for x in range(10+2): ## end before 
    print(x,end="")

01234567891011

In [79]:
for x in range(90,100+1): ## start with and end before 
    print(x,end=" ")

90 91 92 93 94 95 96 97 98 99 100 

In [85]:
num = int(input("Enter a Number:"))

for x in range(0,num):
    print(x,end="")

012345678910

In [97]:
for x in range(100,1000+1,100): ## Jump
    print(x,end=" ")  

100 200 300 400 500 600 700 800 900 1000 

In [101]:
for i in range(100,0,-10): ## Reverse Formula
    print(i,end=" ")

100 90 80 70 60 50 40 30 20 10 

### **for Loop else**
* If loop completes successfully, else will work
* If loop is unsuccessful, else will not work


In [111]:
for x in "hello world":
         print(x,end="")  
else:
 print("\nthe loop completed successfully")

      

hello world
the loop completed successfully


In [115]:
for x in "helloworld":
    if x == " ":
        print(x,end="")
        break
else:
 print("\nthe loop completed successfully")

      


the loop completed successfully


### **find if a number is divisable by 5 in the list*


In [118]:
mylist = [2,6,5,11,16,17]

for each in mylist:
    if each%5 == 0:
        print(each,"Divisiable Buy 5")
        break

else:
 print("number divisiable not found")

5 Divisiable Bu=y 5


### **Nested Loops**

12345
12345
12345
12345
12345
 

In [124]:


for x in range(5): ## first giving Row
    for y in range(1,6):
        print(y,end="")
        y += 1

    x += 1
    print()

12345
12345
12345
12345
12345


In [126]:
for x in ["yellow","green","Red"]:
      print("---------------")
      for y in["cat","dog","car","house"]:
          print(x,y)  
      print("---------------")

print("loop over") 

---------------
yellow cat
yellow dog
yellow car
yellow house
---------------
---------------
green cat
green dog
green car
green house
---------------
---------------
Red cat
Red dog
Red car
Red house
---------------
loop over


In [1]:
row = int(input("Enter a Number: "))
column = int(input("Enter a Number: "))
symbol = (input("Enter a Number: "))


a = 1

while a<=row: ## row
    b = 1
    while b<=column: ## column
        print(symbol,end="")
        b += 1
    
    a += 1
    print()


*****
*****
*****
*****
*****


### **7.User defined functions**

A function/method is a block of code which only runs when it is called.
You can pass data, known as parameter/arguments into a function. A function can return data as a result.

syntax

*  **def \<function_name>\ ():**
* code resueablity


#### * fuction overloading 

----------------------------------------------------------------

### *Keyword*

### * def
### * return

In [ ]:
def happy(): # fuction creation
    print(" i\' love python")

happy() # function calling
happy()
happy()


 i' love python
 i' love python
 i' love python


In [2]:
def happy(): 
    print(" i\' love python")  
    print("code reusabality") 



def  sad():
    print(" i\' love GenAi")  
    print("code reusabality") 

happy()
print()
sad()
print()
happy()



 i' love python
code reusabality

 i' love GenAi
code reusabality

 i' love python
code reusabality


In [5]:
def david():
    pass

### **Function with paramater**

In [ ]:
def myfunc(): ## local varibales
    x = 100
    y = 300
    print("sum  of x & y:",x+y)

myfunc()

sum  of x & y: 400


In [21]:
def myfunc(x,y): ## parameter(varibales) - not permenat temporay
   print("x = " ,  x)
   print("y = ", y)
   print("sum  of x & y:",x+y) ## not saves in memory location and its local variales

myfunc(20,30) ## Arugments
print()

myfunc(100,300)
print()
myfunc(500,0)


x =  20
y =  30
sum  of x & y: 50

x =  100
y =  300
sum  of x & y: 400

x =  500
y =  0
sum  of x & y: 500


In [ ]:
a =  100 ## this global varibales

### **func with defalut paramater**

In [26]:
def myfunc(x=0,y=0): # x ,y =>parameters
   print("x = " ,  x)
   print("y = ", y)
   print("sum  of x & y:",x+y) 

myfunc()
print()
myfunc(20,30)
myfunc(100)
print()
myfunc(y=1000) ## value assign
myfunc("0",y="david")


x =  0
y =  0
sum  of x & y: 0

x =  20
y =  30
sum  of x & y: 50
x =  100
y =  0
sum  of x & y: 100

x =  0
y =  1000
sum  of x & y: 1000
x =  0
y =  david
sum  of x & y: 0david


### **python func with return key**

In [30]:
def myfunc(x=0,y=0):
    print("x = " ,  x)
    print("y = ", y)
    print("sum  of x & y:",x+y) 

ans = myfunc(100,200)
print(ans)
    

x =  100
y =  200
sum  of x & y: 300
None


In [ ]:
a = 100
b = 100
c = a+b 

print(c)

200


In [35]:
def myfunc(x=0,y=0):
    print("x = " ,  x)
    print("y = ", y)
    return x + y

ans = myfunc(100,200)
print("ans:",ans)

print(myfunc(400,400))
    

x =  100
y =  200
ans: 300
x =  400
y =  400
800


In [ ]:
def myadd(x=0,y=0):
    return x+y
def mysub(x=0,y=0):
    return x-y
#             80      20    = 60
ans  =  myadd(50,30)-mysub(30,10)
print(ans)

60


In [4]:
def myadd(x=0,y=0):
    return x+y
def mysub(x=0,y=0):
    return x-y
#             80      20    = 60
ans  =  myadd(50,30)-mysub(30,10)
print(ans)

60


In [ ]:
def myadd(x=0,y=0):
    return x+y

def mysub(x=0,y=0):
    return x-y

#       70        100           30
ans = mysub(myadd(70,30),mysub(50,20))
print("ans",ans)

ans 70


In [ ]:
mylist1 = [1,2,3,4,5] #1*1=1*2=2*3=6*4=24*5=120
mylist2 = [1,2,3,4,5,6]
mylist3 = [1,2,3,4,5,6]

In [16]:

mylist1 = [1,2,3,4,5]

x = 1

for i in mylist1:
    x = x*i
print(x)


120


In [ ]:
## factorial Funcation Automation

mylist1 = [1,2,3,4,5]
mylist2 = [1,2,3,4,5,6]
mylist3 = [1,2,3,4,5,6]


def my_factorial(mylist):
    x = 1

    for i in mylist:
       x = x*i
    return x 

ans  = my_factorial(mylist1)
print(ans)

print(my_factorial(mylist2))
print(my_factorial([1,2,3,4,5,6,7]))



120
720
5040


### **Function with  overloading**

In [ ]:
def myfunc(x):   ##function overrides
    return x

def myfunc(x,y):
    return x + y  ##function overrides

def  myfunc(x,y,z):
    return x+y+z


print(myfunc(100,30,40))


In [ ]:
x = 100
x = 200

print(x)

### **8.String Function**
### *ARBITRARY ARGUMENTS*

If you do not know how many arguments that will be passed into your function, add * before the parameter name in the function definition. This way the function will receive a tuple of arguments, and can access the items accordingly.

* Types of arguments

* *args - argument stores in tuple

* **kwargs - keyworded argument stores in dictionary

In [38]:
## *args we can pass mutiple aruments as datatype

def myfunc(*x):
    print(x,type(x))


myfunc()
myfunc("hai")
myfunc(True)

myfunc((160,"hi",2,55.5,True))

() <class 'tuple'>
('hai',) <class 'tuple'>
(True,) <class 'tuple'>
((160, 'hi', 2, 55.5, True),) <class 'tuple'>


### **string fuctions**

In [ ]:
def myfunc(x=0,y=0):
     print("x =",x)
     print("y=",y)
     return x+y  


ans = myfunc(20,30)
print(ans)

###  ** kwargs - keyworded argument stores in dictionary

In [43]:
def myfunc(**x):
    print(x,type(x))


myfunc(Name ="david",age=27,city="chennai")

{'Name': 'david', 'age': 27, 'city': 'chennai'} <class 'dict'>


### **python string**

* Strings are represented in the form of arrays in Python they are either represented by " " or ' '

str = "HELLO"

##### **5. Python String functions**

##### isalpha() – Checks alphabets
##### isdigit() – Checks digits
##### isalnum() – Checks alphabets + digits
##### islower() – Checks lowercase
##### isupper() – Checks uppercase
##### lower() – Converts to lowercase
##### upper() – Converts to uppercase
##### title() – Capitalizes each word
##### swapcase() – Swaps upper/lower case
##### split() – Splits string into a list
##### join() – Joins elements into a string
##### replace() – Replaces a value
##### count() – Counts occurrences
##### find() / index() – Finds position
##### format() – Formats a string

### **Python String functions**

#### String Indexing

### Positive Index

```text
 0 1 2 3 4 5 6 7 8 9
 H E L L O W O R L D

In [23]:
a = "helloworld"

print(a)    # prints whole string
print(a[0])  #acess the postive Index
print(a[-2])  # acess the negative index
print(a[0:5])  # start index: end before index (1 to 8)
print(a[0:-1])  # start index: end before index (3 to (-2))
print(a[5:])    # start index: till end of string
print(a[:5])    # from  start to string:  end before string
print(a[:])     # from start string to end string
print(a[0:10:2]) # start index: end before index: jump(0 to 9)
print(a[::2])   # from start of string : till end of string : jump 2
print(a[::5])   #from start to string : till end of string : jump
print(a[::2])  # to reverse a string use jump as(-1)



helloworld
h
l
hello
helloworl
world
hello
helloworld
hlool
hlool
hw
hlool


### **String with operators**

In [34]:
print("hello"+"world")
print("hello\n" *5)

helloworld
hello
hello
hello
hello
hello



In [38]:
print("******\n"*5)

******
******
******
******
******



In [42]:
for i in range(1,5):
    print("*"*i)

*
**
***
****


### **String Functions**

### **isalpha()**

In [ ]:
# isalpha() - strictly check whether is alphabet(1,2,3)
# "hell123world" - if there number makes False
# "hello world" -  if there is space also makes False
# "helloworld"  - only if char alphabet without space takes as True

a = "helloworld"
x = a.isalpha()

print(a,type(a))
print(x,type(x))

helloworld <class 'str'>
True <class 'bool'>


In [52]:
a = "hello world"
print(a.isalpha())

print("david".isalpha())

False
True


### **isspace()**

In [1]:
def reverseStr(str):
    count = 0
    for each in str:
        if each.isspace():
           count += 1
         
    return count

ans = reverseStr("David Raj")
print(ans)


1


### **replace()**

In [3]:
name = input("Enter your name: ")

if name.replace(" ","").isalpha():
    print("vaild name")
else:
    print("Invaild name")


vaild name


In [59]:
a = "Helloworld "
x = a.isalpha()
print(x)

False


### **isdigit()**

In [ ]:
# isdiigit() - strictly check whether is alphabet(chars and spaces)
# "1234er" - if there char makes False
# "hello world" -  if there is space also makes False
# "1234"  - only if number without space takes as True

a = "123456 "
x = a.isdigit()
print(x)

False


In [70]:
age  = input("Enter Your Age:  ")

if age.isdigit():
    age = int(age)
    print("vaild age")
else:
    print("invaild age")

invaild age


### **isalnum()**

In [71]:
# isalnum() - strictly checks whether the string contains only alphabets and numbers
# "1234er" - alphabets and numbers are allowed, so True
# "hello world" - if there is a space, it makes False
# "hello@123" - if there is a special character, it makes False
# "Hello123" - only alphabets and numbers without spaces or special characters takes as True

a =  "1234Helloworld"
x = a.isalnum()
print(x)

True


In [73]:
username = input("Enter your username: ")


if username.isalnum():
    print("vaild username")
else:
    print("invaild username")

invaild username


### **islower()**

In [79]:
## islower()

# islower() - strictly checks whether the letters in the string are lowercase
# "hello" - all letters are lowercase, so True
# "Hello" - if there is an uppercase letter, it makes False
# "HELLO" - all letters are uppercase, so False
# "hello world" - lowercase letters with a space are still True
# "hello123" - lowercase letters with numbers are still True
# "hello@123" - lowercase letters with special characters are still True


a = "hello 1234 wold"
x = a.islower()
print(x)

True


In [84]:
message = input("Enter your message: ")

if message.islower():
    print("Lowercase meassage")
else:
    print("not lowercase message")


Lowercase meassage


### **isupper()**

In [82]:
a = "HELLO WORLD"
x = a.isupper()
print(x)

True


In [ ]:
message = input("Enter your message: ")

if message.isupper():
    print("uppercase meassage")
else:
    print("not uppercase message")


Lowercase meassage


### **upper()**

In [4]:
### upper()
# converts a string to upper case

a =  "1224Hello WORld"
x = a.upper()

print(x)

1224HELLO WORLD


### **endswith()**

In [5]:
## ends with()

a = "ABCD Jeevan"
x = a.endswith("n")
print(x)

True


In [6]:
## ends with()

a = "ABCD Jeevan.txt"
x = a.endswith(".txt")
print(x)

True


In [7]:
mylist = ["new.txt","new.py","hello.py","my.txt"]

for each in mylist:
    if each.endswith(".py"):
        print(each)

new.py
hello.py


In [8]:
files = [
    "resume.pdf",
    "photo.jpg",
    "certificate.pdf",
    "profile.png",
    "offer-letter.pdf",
    "notes.txt"
]


for each in files:
    if each.endswith(".pdf"):
        print("PDF file:",each)
    

PDF file: resume.pdf
PDF file: certificate.pdf
PDF file: offer-letter.pdf


### **startswith()**

In [ ]:
mylist = ["new.txt","new.py","hello.py","my.txt"]

for each in mylist:
    if each.startswith("n"):
        print(each)

new.txt
new.py


### **title()**

In [10]:
###  title 
## coverts the First Character of Each Word to Uppercase

a = "HELLO WORLD Its a Great Earth"
x = a.title()
print(x)

Hello World Its A Great Earth


### **capitalize()**

In [11]:
### capitalize()
##  first letter converts to Upper case 

a = "hello WORLD"
x = a.capitalize()

print(x)

Hello world


### **swapcase()**


In [12]:
### swapcase()

## "hello" => change to lower to upper case
## "HELLO" => changess t0 Upper to lower case

a = "Hello sun Its a Sunny day"
x =  a.swapcase()

print(x)


hELLO SUN iTS A sUNNY DAY


In [13]:
name = input("Enter your name: ")

title = name.title()
print("Title:",title)
capital =  name.capitalize()
print("Capitalize:",capital)
swap = name.swapcase()
print("Swapcase:",swap)

Title: Mmj
Capitalize: Mmj
Swapcase: MMJ


### **split()**

In [14]:
## split() - splits the String at the specified separator, and returns a list

a = "Hello sun Its a Sunny day"
x =  a.split()

print(x)

['Hello', 'sun', 'Its', 'a', 'Sunny', 'day']


In [15]:
chat =  "What is Python programming"

separte = chat.split()
print(separte)

['What', 'is', 'Python', 'programming']


In [16]:
question = input("Enter your question: ")

words = question.split()

for each in words:
    print(each)

In [17]:
a = "onomatopoeiao"
x =  a.split("o")
print(x)

['', 'n', 'mat', 'p', 'eia', '']


### **join()**


In [18]:
### join()


a = ["Hello","world","I\"am","David"]

x = " ".join(a)
print(x)


Hello world I"am David


In [19]:
print(" ".join(["David", "yseu"," rahi"]))

David yseu  rahi


### **interview question**

In [ ]:
text = "This is a String"  # string is a This

split_txt = text.split(" ")
reverse = split_txt[::-1]
join_txt = " ".join(reverse)

print(join_txt)


String a is This


In [4]:
mylist  = ("Hello","world","I\"am","david")
print(" ".join(mylist))

Hello world I"am david


### **replace()**


In [6]:
### replace()

a =  "The quick brown for jumps over the lazy dog idiom"
x = a.replace(" ","")
print(x)

Thequickbrownforjumpsoverthelazydogidiom


In [10]:
### replace()

a =  "The quick brown fox jumps over the lazy dog idiom"
x = a.replace("fox","cat")
print(x)

The quick brown cat jumps over the lazy dog idiom


### **replace**

In [11]:
prompt = "I want to learn Java programming for GenAI"

Replaced = prompt.replace("Java","Python")
print(Replaced)

I want to learn Python programming for GenAI


In [12]:
message = "I want to contact SUPPORT"

Replaced = message.replace("SUPPORT","Support")
print(Replaced)

I want to contact Support


### **count()**

In [ ]:
### count()

a =  "The quick brown fox jumps over the lazy dog idiom"

x = a.count("o")
print(x)

5


In [16]:
message = "I need help with my order. My order is delayed and I want to track my order."
x = message.count("order")
print("The word 'order' appears:",x)

The word 'order' appears: 3


### **find()**

In [ ]:
## find()  - find check by index  (1 attempt)
##          - if not found shows -1

# / index() - find check by index

a = "The quick queen"

x = a.find("z")
print(x)

-1


### **index()**

In [ ]:
#index()

a = "The quick queen"

x = a.index("i") #valueError: substring Not found
print(x)

6


In [31]:
message = "I have a problem with my order"

result = message.find("order")
print(result)

25


### **format()**


In [ ]:
## format()
## format()- similar to fill in the blanks {} is used to act as Dash

name = "honey"
print("my name is",name)

my name is honey


In [43]:
## format()

name = "honey"
print("my name is {}".format(name))

my name is honey


In [49]:
name = "david"
city = "chennai"
country = "india"

print("Hello I\"am", name, "I\"am from", city, sep=" ")

print("Hello I\"am {}, I\"am from {}".format(name, city))

Hello I"am david I"am from chennai
Hello I"am david, I"am from chennai


In [50]:
quality = 3
item_no  = 300
price = 667

myorder =  "I want {0} pieces of items No {1} for {2} doller"
print(myorder.format(quality,item_no,price))

I want 3 pieces of items No 300 for 667 doller


In [51]:
myorder = "i want {carname} it is {model}"

print(myorder.format(carname = "ford", model = "mustag"))

i want ford it is mustag


In [7]:
text1 = "my name is {name}, I am {age}".format(name="david",age=15)
text2 = "my name is {}, I am {}".format("jeevan",15)
text3 = "My name is {}, my name is {}".format("jeevan",15)

print(text1)
print(text2)
print(text3)

my name is david, I am 15
my name is jeevan, I am 15
My name is jeevan, my name is 15


In [13]:
myname =  "my is name {name},and my I am {age} years old.".format(name="david",age=27)
print(myname)

my is name david,and my I am 27 years old.


In [16]:
student_detailas = "my name is {},I am Learning {} in {}".format("david","GenAi","chnenai")
print(student_detailas)

my name is david,I am Learning GenAi in chnenai


## **string formatting**

* %d - integer
* %s - string
* %f - float with default 6 decimal places # default 6
* %.2f - float with two decimal places # minumum 2
* %10f - float with ten decimal places # maximum 10


In [17]:
name  = "david"
age = 27
height = 60.15

print("%s is %d year old and height is %.2f" %(name,age,height))

david is 27 year old and height is 60.15


In [18]:
length = 123.5050
wigth = 150.5000
height = 7050.1500

print("length=%f,wigth=%.2f,height=%.8f"%(length,wigth,height))

length=123.505000,wigth=150.50,height=7050.15000000


In [20]:
name = "david"
print("my name is %s"%(name))

my name is david


In [21]:
age = 27
print("my age is %d"%(age))

my age is 27


In [22]:
price = 99.5
print("the price is %f"%(price))

the price is 99.500000


In [23]:
salary = 45000.56789

print("my salary is %.2f" %(salary))

my salary is 45000.57


In [27]:
name = "David"
age = 27
percentage = 85.6789

print("name: %s"%(name))
print("age: %d"%(age))
print("percentage: %.2f"%(percentage))

name: David
age: 27
percentage: 85.68


In [32]:
mystr = "racecar"
revstr = mystr[::-1]
print(revstr)

if mystr == revstr:
    print("True")
else:
    print("False")

racecar
True


In [47]:
mystr = "MalaYa lAm"
newstr = mystr.lower().replace(" ","")
revstr = newstr[::-1]
print(revstr)

if  newstr == revstr:
    print("True")
else:
    print("False")

malayalam
True


In [49]:
mystr = "sir , I demand,I am maid named Iris"
newstr = mystr.lower().replace(" ","").replace(" ","")
revstr = newstr[::-1]
print(revstr)

if  newstr == revstr:
    print("True")
else:
    print("False")

siridemandiammai,dnamedi,ris
False


In [ ]:
mystr = "Red roses run ni risk. Sir, on Nurse's Order"
newstr = ''

for each in mystr:
    if each.isalpha():
        newstr += each.lower()
print(newstr)

if newstr == newstr[::-1]:
    print(True)
else:
    print(False)

redrosesrunnirisksironnursesorder
False


In [ ]:
newstr = ''

for each in mystr:
    if each.isalpha():
        newstr += each.lower()
print(newstr)

if newstr == newstr[::-1]:
    print(True)
else:
    print(False)

In [63]:
def check_palindrome(mystr):
    newstr =''

    for each in mystr:
        if each.isalpha():
            newstr += each.lower()
    print(newstr)

    if newstr == newstr[::-1]:
        return True
    else:
        return False


ans =  check_palindrome("MalaYa lAm")
print(ans)

print(check_palindrome("racecar"))
print(check_palindrome("Red roses run ni risk. Sir, on Nurse's Order"))
print(check_palindrome("sir , I demand,I am maid named Iris"))

malayalam
True
racecar
True
redrosesrunnirisksironnursesorder
False
siridemandiammaidnamediris
True


### dictionaries

In [70]:
student = {
    "name":"david",
    "age" :27,
    "city": "chennai"
}


def display_details(student):
    print(f"name:",student["name"])
    print(f"age:",student["age"])
    print(f"city:",student["city"])

display_details(student)



name: david
age: 27
city: chennai


In [74]:
employee = {
    "id": 101,
    "name": "David",
    "department": "AI",
    "salary": 45000
}

def employee_details(employee):
    print(f"id:",employee["id"])
    print(f"name:",employee["name"])
    print(f"department:",employee["department"])
    print(f"salary:",employee["salary"])

employee_details(employee)

id: 101
name: David
department: AI
salary: 45000


### *JSON stands for JavaScript Object Notation*

### **python dictionary Functions**

# DICTIONARY FUNCTIONS

| Dictionary Function | Description |
|---|---|
| `keys()` | Returns a list containing the dictionary's keys |
| `values()` | Returns a list of all the values in the dictionary |
| `items()` | Returns a list containing a tuple for each key-value pair |
| `pop(key)` | Removes the element with the specified key |
| `popitem()` | Removes the last inserted key-value pair |
| `clear()` | Removes all the elements from the dictionary |
| `update()` | Updates the dictionary with the specified key-value pairs |
| `get()` | Returns the value of the specified key |
| `copy()` | Returns a copy of the dictionary |

In [ ]:
mydict = {
    "name":"david",
    "age":27,
    "city":"chennai",
     500 : 50.55,
    "name":"Raj"
}

print(mydict,type(mydict))
print(mydict["age"])
mydict["age"] = 28
mydict["course"] ="GenAi" # add elements
print(mydict)

{'name': 'Raj', 'age': 27, 'city': 'chennai', 500: 50.55} <class 'dict'>
27
{'name': 'Raj', 'age': 28, 'city': 'chennai', 500: 50.55, 'course': 'GenAi'}


### **keys()**

In [31]:
## keys()

mydict = {
    "name":"david",
    "age":27,
    "city":"chennai",
     500 : 50.55,
    "name":"Raj"
}

print(mydict.keys()) # key prints
print(list(mydict.keys())) ## type casting to list
print()

# ans = list(mydict.keys())
# ans1= list(mydict.values())
# print(ans[0] ,":",ans1[0])
# print(ans[1] ,":",ans1[1])
# print(ans[2] ,":",ans1[2])
# print(ans[3] ,":",ans1[3])



for each in mydict.keys():
    print(each)

dict_keys(['name', 'age', 'city', 500])
['name', 'age', 'city', 500]

name
age
city
500


In [39]:
employee_details = {
    "employee_id":101,
    "name":"david",
    "department":"GenAi",
    "salary": 22000,
    "city":"chennai"
}

print(employee_details)
print(list(mydict.keys()))

for each in employee_details.keys():
    print(each)

{'employee_id': 101, 'name': 'david', 'department': 'GenAi', 'salary': 22000, 'city': 'chennai'}
['name', 'age', 'city', 500]
employee_id
name
department
salary
city


### **values()**


In [42]:
## values()
mydict = {
    "name":"david",
    "age":27,
    "city":"chennai",
     500 : 50.55,
    "name":"Raj"
}

print(mydict.values())
print(list(mydict.values()))

print()

for each in mydict.values():
    print(each)

dict_values(['Raj', 27, 'chennai', 50.55])
['Raj', 27, 'chennai', 50.55]

Raj
27
chennai
50.55


In [16]:
employee = {
    "employee_id": 101,
    "name": "David",
    "department": "GenAI",
    "salary": 22000,
    "city": "Chennai"
    
}

# print(employee.keys())
# print(employee.values())
# print(list(employee.keys()))
# print(list(employee.values()))


# for each in employee.values():
#     print(each)

ans = list(employee.keys())
ans1 = list(employee.values())

for i in range(len(ans)):
    print(ans[i],":",ans1[i])




employee_id : 101
name : David
department : GenAI
salary : 22000
city : Chennai


### **items()**


In [19]:
### items()

employee = {
    "employee_id": 101,
    "name": "David",
    "department": "GenAI",
    "salary": 22000,
    "city": "Chennai"
    
}

print(list(employee.items()))

print()

for each in employee.items():
    print(each)


[('employee_id', 101), ('name', 'David'), ('department', 'GenAI'), ('salary', 22000), ('city', 'Chennai')]

('employee_id', 101)
('name', 'David')
('department', 'GenAI')
('salary', 22000)
('city', 'Chennai')


### **Nested Dictionary**

In [ ]:
mydict = {
    "Name":"David",
    "city":['chennai',"Trichy","Mumbai"],
    "Year":(2000,2001,2010,2026),
}

print(mydict)
print(mydict.keys())
print(mydict.get("Name"))
print(mydict.get("city")[0])
print(mydict.get("city")[1])

print("-------------------")

# print(mydict.get("yearss")) ##if i call unknown value prints None
# print(mydict.get("city")[2][2:])
# print(mydict.get("city")[2:])

{'Name': 'David', 'city': ['chennai', 'Trichy', 'Mumbai'], 'Year': (2000, 2001, 2010, 2026)}
dict_keys(['Name', 'city', 'Year'])
David
chennai
Trichy
-------------------
['Mumbai']


In [36]:
x = {
    "menu": {
        "id": "file",
        "value": "File",
        "popup": {
            "menuitem": [
                {"value": "New", "onclick": "CreateNewDoc()"},
                {"value": "Open", "onclick": "OpenDoc()"},
                {"value": "Close", "onclick": "CloseDoc()"}
            ]
        }
    }
}

### **Charcater Frequency**

In [4]:
# input  = "hello WORLD"
# output = {'h':1,'e':1,'l':3,'o':2,'w':1,'r':1,'d':1}

In [16]:
mystr = "hello WORLD"
newstr = ""

for each in mystr:
    if each.isalpha():
        newstr =newstr+ each.lower()
print (newstr)


mydict = {}

for each_1 in newstr:
    if each_1 not in mydict.keys():
        mydict[each_1] = 1

    else:
        mydict[each_1] += 1

print(mydict)

 

helloworld
{'h': 1, 'e': 1, 'l': 3, 'o': 2, 'w': 1, 'r': 1, 'd': 1}


#### **Python standared functions**

### **len()**


In [ ]:
##len()

x = [1,2,3,4,"Hello",True]
y = "hello world"

print(len(x))  ##item by item takes
print(len(y))  ##letter my letter takes



6
11


### **max()**
### **min()**
### **sum()**

In [24]:
# max()
# min()
# sum()

x = [3, 6, 7, 0, -1, -5, 8, 9, -2, 10, 20, 100, -9, -7]

print(x)
print(min(x))
print(max(x))
print(sum(x))

[3, 6, 7, 0, -1, -5, 8, 9, -2, 10, 20, 100, -9, -7]
-9
100
139


### **abs**

In [29]:
a = 25
b = -25
b1 = abs(b)
a1 = abs(a)

print(a)
print(b)
print(b1)
print(a1)

25
-25
25
25


### **pow()**

In [ ]:
print(10**3) ##1000
print(pow(10,3)) ##1000



1000
1000


### **round**

In [ ]:
x = 50.55
y = 40.55
x1 = int(x)
print(x1)

print(round(x)) #round value

50
51


### **slice**

In [41]:
x = [0,1,2,3,4,5,6,7,8,9,10]

print(x)
print(x[3:9])

print(x[slice(3,9)])

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[3, 4, 5, 6, 7, 8]
[3, 4, 5, 6, 7, 8]


### **sorted**

In [46]:
x = [-4,4,5,-8,-7,4,-3,9,10,2,4,7,-1]

print(sorted(x))

x.sort()
x.reverse()
print(x)

[-8, -7, -4, -3, -1, 2, 4, 4, 4, 5, 7, 9, 10]
[10, 9, 7, 5, 4, 4, 4, 2, -1, -3, -4, -7, -8]


In [50]:
x = [-4,4,5,-8,-7,4,-3,9,10,2,4,7,-1]

print(sorted(x))
print(sorted(x,reverse=True))

[-8, -7, -4, -3, -1, 2, 4, 4, 4, 5, 7, 9, 10]
[10, 9, 7, 5, 4, 4, 4, 2, -1, -3, -4, -7, -8]


### **char()**
### **ord()**

In [52]:
print(ord("a"))
print(chr(97))

97
a


In [55]:
for i in range(33,127):
    print(chr(i),end=" ")

! " # $ % & ' ( ) * + , - . / 0 1 2 3 4 5 6 7 8 9 : ; < = > ? @ A B C D E F G H I J K L M N O P Q R S T U V W X Y Z [ \ ] ^ _ ` a b c d e f g h i j k l m n o p q r s t u v w x y z { | } ~ 

### **enumerate()**

In [9]:
x  = [4,5,6,7,8,9,5,4,3,2,1]


print(x.index(5))

1


In [10]:
#index()

a = "The quick queen"

x = a.index("i") #valueError: substring Not found
print(x)

6


In [3]:
x  = [4,5,6,7,8,9,5,4,3,2,1]

print(list(enumerate(x))) ## tuple pair


[(0, 4), (1, 5), (2, 6), (3, 7), (4, 8), (5, 9), (6, 5), (7, 4), (8, 3), (9, 2), (10, 1)]


### **enumerate()**

* Multiple references → we can use enumerate() to get both the index and value.
* For a single reference → we can use index() to find the position of a value.
* enumerate() returns index-value pairs as tuples.
* nameless functions

In [18]:
x  = [4,5,6,7,8,9,5,4,3,2,1]

mylist = list(enumerate(x)) ## tuple pair


for each in mylist:
    print(each)


(0, 4)
(1, 5)
(2, 6)
(3, 7)
(4, 8)
(5, 9)
(6, 5)
(7, 4)
(8, 3)
(9, 2)
(10, 1)


In [23]:
x  = [4,5,6,7,8,9,5,4,3,2,1,2]
search =2

mylist = list(enumerate(x)) ## tuple pair


for each in mylist:
    if each[1] == search:
       print(f"the element found at index = {each[0]}")
    


the element found at index = 9
the element found at index = 11


In [24]:
x  = "hello"
search ="l"

mylist = list(enumerate(x)) ## tuple pair


for each in mylist:
    if each[1] == search:
       print(f"the element found at index = {each[0]}")
    


the element found at index = 2
the element found at index = 3


In [29]:
x = [4, 7, 2, 7, 9, 7]

search = 7
counter = 0
mylist = list(enumerate(x)) ## tuple pair


for each in mylist:
    if each[1] == search:
       print(f"First occurrence of 7 = index {each[0]}")
       counter += 1
  
    


First occurrence of 7 = index 1
First occurrence of 7 = index 3
First occurrence of 7 = index 5


### **LAMBDA FUNCTION**

* • A lambda function is a small anonymous function.
* • A lambda function can take any number of arguments, but can only have one expression.
* name less functions

Syntax:

*  funcname = lambda arguments : expression

In [4]:
new_func = lambda x,y : x*y
print(new_func(20,30))

ans1 = new_func(50,6)
print(ans1)


ans2 = new_func(70,8)
print(ans2)

600
300
560


### **map()**

* map(Fuction_name,Iterable)
* map checks each value by value

In [ ]:
# input = [1,2,3,4,5]
# output = [1,2,9,16,25]

old_list = [1,2,3,4,5]
new_list = []

for each in old_list:
    new_list.append(each*each*each)

print(new_list)


[1, 8, 27, 64, 125]


In [8]:
old_list = [1,2,3,4,5]

def square(x):
    return x*x

def cube(n):
    return n*n*n

ans1 = list(map(square,old_list))
print(ans1)

ans2 = list(map(cube,old_list))
print(ans2)


[1, 4, 9, 16, 25]
[1, 8, 27, 64, 125]


In [13]:
square =  lambda x:x*x
cube = lambda n:n*n*n

old_list = [1,2,3,4,5]

ans1 = list(map(square,old_list))
ans2 = list(map(cube,old_list))

print(ans1)
print(ans2)

[1, 4, 9, 16, 25]
[1, 8, 27, 64, 125]


In [18]:
print(list(map(lambda x : x*x,[1,2,3,4,5])))
print(list(map(lambda x : x*x*x,[1,2,3,4,5])))

[1, 4, 9, 16, 25]
[1, 8, 27, 64, 125]


### **filter**


* filter(Fuction_name,Iterable)
* filter by condition given 

In [9]:
## must try insert and extend

input =  [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,188,19,20]

new_list = []

for each in input:
    if each%3 == 0:
        new_list.insert(0,each)
        
print(new_list)


[15, 12, 9, 6, 3]


In [10]:
print(list(filter(lambda x : x%3 ==0,[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,188,19,20])))

[3, 6, 9, 12, 15]


### **zip()**

In [37]:
colour  = ["Black","Red","Green","blue"]
vehicles  = ["bike","car","van","bus"]

ans = list(zip(colour,vehicles))
print(ans)

print()

for each in ans:
    print(f"{each[0]}\t{each[1]}")

[('Black', 'bike'), ('Red', 'car'), ('Green', 'van'), ('blue', 'bus')]

Black	bike
Red	car
Green	van
blue	bus


### **GENERATOR FUNCTION**

A generator-function is defined like a normal function, but whenever it needs to generate a value, it does so with the yield keyword rather than return. If the body of a def contains yield, the function automatically becomes a generator function.

###  **yield** - keyword

In [40]:
def myfunc():
    x = 50
    y = 50
    z =  x+y
    return z

ans = myfunc()
print(ans)

100


In [11]:
def myfunc():
    yield 100
    yield 200
    yield 300

ans = list(myfunc())
print(ans)

[100, 200, 300]


In [2]:
def myfunc():
    yield 100
    yield "david"
    yield 300.55

for each in myfunc():
    print(each,type(each))

100 <class 'int'>
david <class 'str'>
300.55 <class 'float'>


In [1]:
def myfunc():
    print("This is my myfunc calling 1")
    yield 100 ## directy goes to loop
    print("This is my myfunc calling 1")
    yield 200
    print("This is my myfunc calling 1")
    yield 300
    print("This is my myfunc calling 1")
    yield 400

for each in myfunc():
    print("loop in next iteration")
    print(each)

This is my myfunc calling 1
loop in next iteration
100
This is my myfunc calling 1
loop in next iteration
200
This is my myfunc calling 1
loop in next iteration
300
This is my myfunc calling 1
loop in next iteration
400


### **help**

* q letter for come out from 

In [6]:
#help()

# help()

### **eval()**

In [12]:
eval('print("hello world")')

hello world


In [14]:
expression = "(25 + 15) * 2 - 10 / 5"

result = eval(expression)

print(result)

78.0


In [14]:
# eval("hello world") - syntax error

### **modules**
* import 
* as
* from

In [19]:
import math 

ans  = math.sqrt(10)
print(ans)

3.1622776601683795


### **import**

In [20]:
import  math

ans = math.factorial(5)
print(ans)

120


### **as**

In [22]:
#as
import math as m
ans = m.factorial(7)
print(ans)

5040


In [23]:
import pandas as pd
import numpy as np

### **from**

In [28]:
#from

from math import sqrt,factorial,log

print(sqrt(8))
print(factorial(8))
print(log(8))

2.8284271247461903
40320
2.0794415416798357


### *from new import* *

In [ ]:
#from new import *  # importing all full file access

### **SCOPE**

Scope refers to the coding region from which particular Python object is accessible. Hence one cannot access any particular object from anywhere from the code, the accessing has to be allowed by the scope of the object.

Types of scope

* Local scope
* Global scope


In [9]:
x = 100
y = 200

print(f"x is Global scope = {x}")
print(f"y is Global scope = {y}")

print()

def myfunc(x=0,y=0):
    print(f"x is Global scope = {x}")
    print(f"y is Global scope = {y}")

myfunc()
print()
myfunc(1000,500)
print()
myfunc()

x is Global scope = 100
y is Global scope = 200

x is Global scope = 0
y is Global scope = 0

x is Global scope = 1000
y is Global scope = 500

x is Global scope = 0
y is Global scope = 0


* ### if there no value in local it takes value from global ###

In [4]:
x = 10

def myfunc():
    y = 200
    print(x,y)

print(x)

myfunc()

10
10 200


* ### if the value alreday in local does not take from global value ###

In [7]:
x = 10

def  myfunc():
    x = 1000
    y = 200

    print(y,x)

print(x)
myfunc()

10
200 1000


In [12]:
x = 10

def myfunc():
    x = 1000
   
    global y

    y = 500

    print(x,y)

print(x)
myfunc()
print(y)

10
1000 500
500


### **RECURSION**

A function is called within the same function is called Recursion.

Recursion is a common mathematical and programming concept. It means that a function calls itself. This has the benefit of meaning that you can loop through data to reach a result.

The developer should be very careful with recursion as it can be quite easy to slip into writing a function which never terminates, or one that uses excess amounts of memory or processor power. However, when written correctly recursion can be a very efficient and mathematically-elegant approach to programming.

In [3]:
def myadd(x,y):
    return x+y
def mysub(x,y):
    return x-y

print(mysub(myadd(10,20),myadd(5,10)))

15


In [8]:
def myfactorial(num):
    pro = 1

    for i in range(1,num+1):
        pro = pro*i

    return pro

ans  = myfactorial(5)
print(ans)

120


In [3]:
def myrec_factorial(num):
    if num == 1:
        return 1
    else:
        return num * myrec_factorial(num-1)
ans = myrec_factorial(5)
print(ans)

120


### **Fibonacci Series Using recursion**

In [7]:
#fibonacci series using Recursion

def recur_fibo(n):
    if n<= 1:
        return n
    else:
        return(recur_fibo(n-1)+recur_fibo(n-2))

nterms = 10

##check if the Number of terms is vaild

if nterms<=0:
    print("please Enter a postive Number")
else:
    print("Fibonacci Sequances:")
    for i in range(nterms):
        print(recur_fibo(i))

Fibonacci Sequances:
0
1
1
2
3
5
8
13
21
34


### **Exception Handling or (Error Handling)**

* complie time error
* intentation error
* Zerodivision error

##### **In Python Two Major Errors**
* syntax error
* logical error

Try … Except … finally

* The try block lets you test a block of code for errors.
* The except block lets you handle the error.
*  finally block lets you execute code, regardless of the result of the try and except blocks.
* The else block works only if try is correct.

| Error | Meaning | Example |
|---|---|---|
| `ZeroDivisionError` | Dividing a number by zero | `10 / 0` |
| `ValueError` | Invalid value for a function | `int("abc")` |
| `TypeError` | Wrong data type used in an operation | `"abc" + 10` |
| `IndexError` | Accessing an out-of-range list index | `my_list[10]` |
| `KeyError` | Accessing a non-existing key in a dictionary | `my_dict["not_found"]` |
| `AttributeError` | Accessing an attribute of an object that does not exist | `None.upper()` |
| `FileNotFoundError` | File not found when opening | `open("abc.txt")` |
| `ImportError` | Trying to import a missing module | `import abc_module` |
| `NameError` | Using a variable that was never defined | `print(xyz)` |
| `IndentationError` | Incorrect indentation | Code not properly indented |
| `SyntaxError` | Writing invalid Python code | `if True print("hi")` |
| `RuntimeError` | Generic error that does not fall into other categories | Rare cases |
| `StopIteration` | Raised by `next()` when an iterator is exhausted | In loops and generators |
| `MemoryError` | Program runs out of memory | Working with huge data |
| `RecursionError` | Too many recursive function calls | A function calls itself endlessly |

In [3]:
a = 10
b = 0

result = a/b  #cashed here
print("Result", result)

ZeroDivisionError: division by zero

In [5]:
print("Welcome to zomoto")
number_of_items = int(input("how many items?"))
total_price =  200*number_of_items
average_price = total_price/number_of_items # Division  by zero
print("Average price per item:",average_price)

Welcome to zomoto


ZeroDivisionError: division by zero

In [1]:
    print("Welcome to zomoto")
    number_of_items = int(input("how many items?"))
    total_price =  200*number_of_items
    average_price = total_price/number_of_items # Division  by zero
    print("Average price per item:",average_price)

Welcome to zomoto


ZeroDivisionError: division by zero

### **else**

In [7]:
try:
    print("Welcome to Zomato")
    number_of_items = int(input("How many items? "))
    total_price = 200 * number_of_items
    average_price = total_price / number_of_items
    print("Average price per item:", average_price)

except ZeroDivisionError:
    print("Cannot divide by zero")

else:
    print("No error found")

finally:
    print("Code always executes")

print("Next code block")

Welcome to Zomato
Cannot divide by zero
Code always executes
Next code block


###  **finallly** 
* ### always works ###

In [5]:
try:
    print("Welcome to zomoto")
    number_of_items = int(input("how many items?"))
    total_price =  200*number_of_items
    average_price = total_price/number_of_items # Division  by zero
    print("Average price per item:",average_price)

except FileNotFoundError:
    print("file not found")

finally:
    print("Code Always excute")

print("next code block")

Welcome to zomoto
Code Always excute


ZeroDivisionError: division by zero

In [ ]:
try:
    print("Welcome to zomoto")
    number_of_items = int(input("how many items?"))
    total_price =  200*number_of_items
    average_price = total_price/number_of_items # Division  by zero
    print("Average price per item:",average_price)

except ZeroDivisionError:
    print("you cannot order 0 items")

print("next code block")

In [2]:
try:
    print("Welcome to zomoto")
    number_of_items = int(input("how many items?"))
    total_price =  200*number_of_items
    average_price = total_price/number_of_items # Division  by zero
    print("Average price per item:",average_price)

except Exception:
    print("you cannot order 0 items")

print("next code block")

Welcome to zomoto
you cannot order 0 items
next code block


In [4]:
try:
    print("Welcome to zomoto")
    number_of_items = int(input("how many items?"))
    total_price =  200*number_of_items
    average_price = total_price/number_of_items # Division  by zero
    print("Average price per item:",average_price)

except Exception as e:
    print(e)

print("next code block")

Welcome to zomoto
invalid literal for int() with base 10: '-='
next code block


In [ ]:
# Syntax error (syntax error)
print("hello world"

_IncompleteInputError: incomplete input (2020447913.py, line 2)

In [ ]:
# Syntax error (name error)
prin("hello world"

_IncompleteInputError: incomplete input (363656573.py, line 2)

In [8]:
## logical error (value Error)

print("Few lines below")

num = int(input("Enter a Number:"))
print(f"Enter Number is:{num}")


print("Few lines below")


Few lines below


ValueError: invalid literal for int() with base 10: 'hf'

### **Error Handling**
* try
* except

In [9]:
try:
  num = int(input("Enter a Number:"))
  print(f"Enter Number is:{num}")
  
except:
  print("Error Handling is done here")

print("this line always Woks")

Error Handling is done here
this line always Woks


In [10]:
while True:
    try:
        num = int(input("Enter a Number:"))
        print(f"Enter Number is:{num}")
        break
    
    except:
        print("Plaese Enter a integer Number eg, -99,100,-662345")

print("this line always Woks")

Plaese Enter a integer Number eg, -99,100,-662345
Plaese Enter a integer Number eg, -99,100,-662345
Plaese Enter a integer Number eg, -99,100,-662345
Plaese Enter a integer Number eg, -99,100,-662345
Enter Number is:9
this line always Woks


### **Exception**

In [11]:
while True:
    try:
        num = int(input("Enter a Number:"))
        print(f"Enter Number is:{num}")
        break
    
    except Exception as e:
        print(e)

print("this line always Woks")

invalid literal for int() with base 10: 'df'
invalid literal for int() with base 10: 'yj0-\\'
Enter Number is:7
this line always Woks


### **ZeroDivisonError**

In [ ]:
print(10/0) ## the vaule cannot divide by zero

ZeroDivisionError: division by zero

In [16]:
print("Few lines Above")

try:
    num1 = int(input("Enter a Dividend number"))
    print(f"The user Entered Dividend Number is = {num1}")

    num2 = int(input("Enter a Divisor number"))
    print(f"The user Entered Divisor Number is = {num2}")

    quotient = num1/num2
    print(f"the Quotient of{num1}/{num2} ={quotient}")

except ValueError:
       print("You gave wrong input type")

except ZeroDivisionError:
    print("Don't divide by zero")

print("Few lines Above")


Few lines Above
You gave wrong input type
Few lines Above


In [15]:
try:
    print(int(input("Enter  Number:")))
except exception as e:
    print(e)

NameError: name 'exception' is not defined

In [17]:
while True:
    try:
        num = int(input("Enter a Number:"))
        print(f"Enter Number is:{num}")
        
    
    except :
        print("Error handling done here")
        print(10/0)

print("this line always Woks")

Enter Number is:20
Error handling done here


ZeroDivisionError: division by zero

### **else**
* **the **else** block works only if try is correct**

In [18]:
print("Few lines Above")

try:
    num1 = int(input("Enter a Dividend number"))
    print(f"The user Entered  Number is = {num1}")

except:
       print("Error Handling is Done Here")
else:
    print("excutes since try has no errors")

print("Few lines Above")


Few lines Above
The user Entered  Number is = 10
excutes since try has no errors
Few lines Above


### **raise - keyword**
* we create the error using rasie keyword

In [21]:
print("Few lines above")
raise ValueError("Hi, i am rising by error message") ##(Error create by ourself)
print("Few lines above")

Few lines above


ValueError: Hi, i am rising by error message

### **Error handling with Raise keyword**

In [26]:
num =  int(input("Enter a number between 10-50:"))

if num<10:
    raise ValueError
elif num>50:
    raise ZeroDivisionError 
else:
    print("Thank you for entering the {num}")

ZeroDivisionError: 

### **Handling error with raise & Except**

In [28]:
while True:
    try:
        num =  int(input("Enter a number between 10-50:"))

        if num<=10:
            raise ValueError
        elif num>=50:
            raise ZeroDivisionError 
        else:
            print(f"Thank you for entering the {num}")
            break

    except ValueError:
        print("Plese Enter a number greater than or equal to 10")
    except ZeroDivisionError:
        print("Please Enter a number lesser than or equal to 50")


Thank you for entering the 16


* **raise creates/triggers the error.**
* **except catches/handles the error.**


In [32]:
while True:
    try:
        num = int(input("Enter a number in between 10-50: "))

        if num < 10:
            raise ValueError

        elif num > 50:
            raise ValueError

        else:
            print(f"Thank you for entering the {num}")
            break

    except ValueError:
        print("Please enter a number between 10 and 50")

Please enter a number between 10 and 50
Please enter a number between 10 and 50
Please enter a number between 10 and 50
Please enter a number between 10 and 50
Please enter a number between 10 and 50
Thank you for entering the 16


In [52]:
vaild_id =  123
id_number = int(input("Enter Id Number:"))

try:
    if id_number != 123:
        raise ValueError("Invaild Id 🚨")

    print(f"ID verified: {vaild_id},Entry allowed")

except  ValueError as e:
    print(e)
    print("Security team will handle the alert.")

Invaild Id 🚨
Security team will handle the alert.


In [57]:
while True:
    try:
        num = int(input("Enter a Number In between 10-50"))
        if num<10:
            raise ValueError ("Please Enter a number greater than 10 or equal to 10")
        elif num>50:
            raise ZeroDivisionError ("Please Enter a number greater than 10 or equal to 10")
        else:
            print(f"Thanks for Entering the = {num}")
            break

    except ValueError as e1:
        print(e1)
    except ZeroDivisionError as e2:
       print(e2)


Please Enter a number greater than 10 or equal to 10
Please Enter a number greater than 10 or equal to 10
Please Enter a number greater than 10 or equal to 10
Thanks for Entering the = 32


In [58]:
try:
    num = 10 / 0

except ZeroDivisionError as e:
    print(e)

division by zero


### **assert keyword**
* if condition false assert will work
* if condition True assert will not work

In [64]:
x = 5
assert x<6

In [62]:
x = 5
assert x<3

AssertionError: 

In [65]:
x =3
assert z>5,"my vaule is lesser than 5"

NameError: name 'z' is not defined

In [66]:
while True:
    try:
        num = int(input("Enter a number:"))

        assert num>=10 and num<=50,"Please Enter Number Between 10-50"
        print(f"Thanks for Entering the {num}")
        break

    except Exception as e:
        print(e)


Please Enter Number Between 10-50
Please Enter Number Between 10-50
Please Enter Number Between 10-50
Please Enter Number Between 10-50
Please Enter Number Between 10-50
Thanks for Entering the 22


In [ ]:
valid_id = 123

print("------- ID VERIFICATION ----------")
print()

while True:
    try:
        id_number = input("Enter ID Number: ")

        assert id_number.isdigit(), "Please enter numbers only."

        if int(id_number) != valid_id:
            raise ValueError(f"Invalid-ID {id_number} 🚨")

        print(f"Thank you, Your-ID {valid_id} verified. Entry allowed.")
        break

    except AssertionError as e:
        print(e)

    except ValueError as e:
        print(e)
        print("Security team has been alerted. 🚨")

-------ID VERIFICATION----------

Invalid-ID134🚨
Security team has been alerted. 🚨
Thank you,Your-ID 123 verified. Entry allowed.


In [86]:
try:
    print("outer block")

    try:
        num = int(input("enter a number:"))
        result = 10/num
        print(result)

    except ValueError:
        print("Inner Error:Invaild input(not a number)")
    except ZeroDivisionError:
        print("Inner ErrorL:Cannot divide by Zero")

except Exception as e:
    print("Outer Error:",e)

outer block
1.0


In [ ]:
print("few lines above")

while True:
    try:
        num1 = int(input("enter a dividend number:"))
        print(f"the user Entered Diviend Number is {num1}")

        

### **Python Regular Experssion**
**************************************************************

* A RegEx, or Regular Expression, is a sequence of characters that forms a search pattern.
* RegEx can be used to check if a string contains the specified search pattern.
* Python has a built-in package called re, which can be used to work with Regular Expressions.

### **^ - start with**

In [1]:
import re

text = "hello world"

a = re.search("^hello",text)
print(a)

if a is not None:
    print("match found")
else:
    print("match not found")

<re.Match object; span=(0, 5), match='hello'>
match found


###  * - Ends with

In [3]:
import re

text = "hello world"

a = re.search("world$",text)
print(a)

if a is not None:
    print("match found")
else:
    print("match not found")

<re.Match object; span=(6, 11), match='world'>
match found


###  **findall()**

In [12]:
import re

text = "hello world helloo"

a = re.findall("hel.*o",text)
print(a)

['hello world helloo']


## REGEX META CHARACTERS

- `[]` → A set of characters
  - Example: `[a-m]`

- `\` → Signals a special sequence / escapes special characters
  - Example: `\d`

- `.` → Any character except newline
  - Example: `he..o`

- `^` → Starts with
  - Example: `^hello`

- `$` → Ends with
  - Example: `world$`

- `*` → Zero or more occurrences
  - Example: `aix*`

- `+` → One or more occurrences
  - Example: `aix+`

- `?` → Zero or one occurrence
  - Example: `aix?`

- `{}` → Exactly the specified number of occurrences
  - Example: `al{2}`

- `|` → Either / OR
  - Example: `falls|stays`

- `()` → Capture and group
  - Example: `(aix)+`

# REGEX : SETS

| Set | Description |
|---|---|
| `[arn]` | Returns a match where one of the specified characters (a or r or n) are present |
| `[a-n]` | Returns a match for any lower case character, alphabetically between a and n |
| `[^arn]` | Returns a match for any character EXCEPT a, r, and n |
| `[0123]` | Returns a match where any of the specified digits (0, 1, 2, or 3) are present |
| `[0-9]` | Returns a match for any digit between 0 and 9 |
| `[0-5][0-9]` | Returns a match for any two-digit numbers from 00 and 59 |
| `[a-zA-Z]` | Returns a match for any character alphabetically between a and z, lower case OR upper case |
| `[+]` | In sets, `+`, `*`, `.`, `|`, `()`, `{}`, `$` has no special meaning, so `[+]` means: return a match for any `+` character in the string |

## REGEX : SPECIAL SEQUENCES

- `\A` → Match if characters are at the beginning of the string
  - Example: `\AThe`

- `\b` → Match at the beginning or end of a word
  - Example: `r"\bain"`
  - Example: `r"ain\b"`

- `\B` → Match when characters are NOT at the beginning or end of a word
  - Example: `r"\Bain"`
  - Example: `r"ain\B"`

- `\d` → Matches digits (0-9)
  - Example: `\d`

- `\D` → Matches characters that are NOT digits
  - Example: `\D`

- `\s` → Matches whitespace characters
  - Example: `\s`

- `\S` → Matches characters that are NOT whitespace
  - Example: `\S`

- `\w` → Matches word characters (a-z, A-Z, 0-9, `_`)
  - Example: `\w`

- `\W` → Matches characters that are NOT word characters
  - Example: `\W`

- `\Z` → Match if characters are at the end of the string
  - Example: `Spain\Z`

## REGEX FUNCTIONS

| Function | Description |
|----------|-------------|
| `findall()` | Returns a list containing all matches |
| `search()` | Returns a Match object if there is a match anywhere in the string |
| `split()` | Returns a list where the string has been split at each match |
| `sub()` | Replaces one or many matches with a string |

In [13]:
import re

text = "I am learning Python"

result = re.search("Python",text)
print(result.group())

Python


In [17]:
import re

text = "I have 25 apples and 10 oranges"

result = re.findall(r"\d+", text)

print(result)

['25', '10']


In [ ]:
import re

text = "David"

result = re.findall(r"[abc]", text)

print(result)

['b', 'a']


In [21]:
import re

text = "David"

result = re.findall(r"[aeiou]", text)

print(result)

['a', 'i']


In [22]:
import re

text = "David Python"

result = re.findall(r"[A-Z]", text)

print(result)

['D', 'P']


In [23]:
import re

text = "David Python"

result = re.findall(r"[a-z]", text)

print(result)

['a', 'v', 'i', 'd', 'y', 't', 'h', 'o', 'n']


In [27]:
import re

text = "My age is 27"

result = re.findall(r"[0-9]", text)

print(result)

['2', '7']


In [28]:
import re

text = "My age is 27"

result = re.findall(r"[0-9]+", text)

print(result)

['27']


In [36]:
import re

text = "David123 python"

result = re.findall(r"[a-zA-Z]+", text)

print(result)

['David', 'python']


In [13]:
import re

myfile = open("C:\\Users\\Yesu\\OneDrive\\Documents\\Desktop\\david.txt")


mydata = myfile.read()


mylist = re.findall(r"[Aa][n][d]",mydata)
print(mylist)

['and', 'and', 'and', 'and', 'And']


## **File Handling**

### Modes (One-Liner Summary):
### **Mode    Description**
* 'r'     Read-only (file must exist)
* 'w'     Write-only (overwrites or creates)
* 'a'     Append-only (adds to end of file)
* 'r+'    Read + write (file must exist)
* 'w+'    Write + read (overwrites or creates)
* 'a+'    Append + read (creates if not exists)
* 'rb'    Read binary
* 'wb'    Write binary
* 'ab'    Append binary

### **1.Write mode**

In [21]:
file = open("myfile.txt","w")

file.write("Welcome to Genai\n")
file.write("i love pythonhi\n")
file.close()

### **2.Read mode**

In [22]:
file = open("myfile.txt","r")
content = file.read()
print("File content:\n", content)
file.close()

File content:
 Welcome to Genai
i love pythonhi



### **3.append mode**

In [ ]:
file = open("myfile.txt","a")
file.write("Adding a newline from Append\n")
file.close() 

In [24]:
file = open("myfile.txt","w")
file.write("Adding a newline from Append\n")
file.close()

### **4.with mode**
* with using we no need close file

In [28]:
with open("myfile.txt","r") as file:
    for line in file:
        print(line.strip())

Adding a newline from Append


### **login user feedback**

In [32]:
feedback = input("enter your feedback:")

with open("myfile.txt","a") as log:
    log.write(feedback + "\n")

print("Thanks for your feedback")

Thanks for your feedback


In [33]:
with open("myfile.txt","r") as file:
    print(file.readline().strip())
    print(file.readline().strip())
    print(file.readline().strip())

Adding a newline from Append
i love GenAi Enginerring
i love python


### **5.readline()**

In [5]:
with open("file.txt") as f:
    while True:
        line = f.readline()

        if not line:
            break
        if "ERROR" in line:
            print("Found error:", line.strip())

Found error: ERROR: User login Not completed


### **Throw and varibale** ***-***
* no need consider anything to print 

In [8]:
with open("file.txt") as f:
    for _ in range(10):
        print(f.readline().strip())

INFO: Application started Sucessfully
ERROR: User login Not completed
INFO: fail to connect database
ftjykui
jtygh
jtyghrdh
dh
jthj
drht6u
final line 10


In [11]:
with open("input.csv","r") as infile, open("output.csv","w") as outfile:
    for line in infile:
        print(line.strip())
        outfile.write(line)

id,name,age
1,david,27
2,sam,28
3,yesu,30


## **os - Modules**

In [26]:
import os

print(os.getcwd())

x = open("myfile.txt","x")
x.close()

c:\Users\Yesu\OneDrive\Documents\Desktop\First_folder\Python_classes


FileExistsError: [Errno 17] File exists: 'myfile.txt'

In [ ]:
import os

print(os.getcwd())

os.chdir("C:\\Users\\Yesu\\OneDrive\\Documents\\Desktop")
print(os.getcwd())



c:\Users\Yesu\OneDrive\Documents\Desktop\First_folder\Python_classes
C:\Users\Yesu\OneDrive\Documents\Desktop


In [ ]:
import os

data  = os.listdir("C:\\Users\\Yesu\\OneDrive\\Documents\\Desktop")


for each in data:
    print(each)


14939 [Converted].ai
advanced-python-with-django.pdf
back.jpg
Blood realtions.png
Buisness card 2.jpg
Buisness card copy.psd
Buisness card.psd
Buisness card1.jpg
Buisness card11.jpg
Buisness card2.psd
Buisness card22.jpg
demo
demo.py.txt
desktop.ini
etryhy.png
Feedback Analyzer
files
First_folder
front.jpg
genai
GenAI_Core_Python_Must_Learn_Roadmap.pdf
GenAI_Engineer_Master_Roadmap_Final_Ordered.pdf
Gen_AI Resume_David.pdf
Gen_AI Syllabus.pdf
Lindin
New folder
Python - Copy (1).pdf
Python Full Stack Developer with Gen AI Training Syllabus in ACTE (1).pdf
Python Programming + Advanced Python for Gen AI (1).pdf
qywehr.png
rbe fd.png
Resume Critiquer
Screenshot 2026-09-05 094935.png
Screenshot 2026-09-18 005851.png
Screenshot 2026-09-19 120632.png
Screenshot 2026-09-19 123745.png
VDB.png
vdf f bg.png
~$AI-Driven Multi-Domain Engineering Assistant - YantraAI.pptx


In [ ]:
import os

data  = os.listdir("C:\\Users\\Yesu\\OneDrive\\Documents\\Desktop")


for each in data:
    if each.startswith("demo"):
        print(each)


demo
demo.py.txt
